Set explicit timeout for http request

lsiepel opened the pull request:
After merging https://github.com/openhab/openhab-addons/pull/15411 it look liked more bindings could have a possible issue with http timeouts. 

So i searched for "request.send(" and inspected all 229 cases  ;-)  For each reqeust that did not have an explicit timeout set, it was added.

Used this basic rule:
If the binding has a timeout config or constant, it was re-used. Otherwise a constant was created and set to 10 seconds.

jlaur commented on bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/IndegoController.java:
```diff
@@ -59,6 +60,7 @@ public class IndegoController {
 
     private static final String BASE_URL = "https://api.indego-cloud.iot.bosch-si.com/api/v1/";
     private static final String CONTENT_TYPE_HEADER = "application/json";
+    private static final int REQUEST_TIMEOUT_MS = 10000;
```
How was this timeout value decided? Please keep in mind that this is a cloud service that might be slow from time to time. I have know idea what would be the best value, so for now just asking the question since the same value is used across all bindings.

jlaur replied:
In some cases the mower is queued over cellular network, and it has to be woken up when asleep. These calls are pretty slow, so I'm pretty sure this timeout must be much higher (or dropped).

daMihe replied:
Yes, i agree. 10 seconds may be to less. The bigger flaw from my point of view is, that's not configurable, but hard coded. Maybe 10 seconds are enough for most requests. But i would not assume this is true for every system and every server. This should be at least configurable using a system property or similar (which makes this change more complex).

lsiepel replied:
There was no scientific research to determine the general timeout for this PR. To me in general 10 second seemed reasonable, but i agree there might be some specific bindings http calls that may act slower.
 (even though i have never seen anything that slow)

Anyway, i'm happy to adjust any binding if needed. I tripled this one to 30 seconds.

jlaur commented on bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/IndegoController.java:
```diff
@@ -119,6 +121,7 @@ protected <T> T getRequest(String path, Class<? extends T> dtoClass)
         int status = 0;
         try {
             Request request = httpClient.newRequest(BASE_URL + path).method(HttpMethod.GET)
+                    .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MICROSECONDS)
```
This method is also used for **longpoll**, so it can't simply be changed to 10 seconds. See:
https://github.com/openhab/openhab-addons/blob/7fb9efc8850236e515e1fbab10eee867620b6b56/bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/IndegoDeviceController.java#L97-L109

and also #13182. You now triggered me to revisit this. 🙂

lsiepel replied:
Should i leave this out, so you can set something up as you proceed in #13182 ?

jlaur replied:
Yes, **longpoll** and synchronous calls over cellular network to a sleeping robot complicates it for this binding. If you can revert changes for this binding, I'll make sure to also lower the default timeout where possible when going back to this topic.

lsiepel replied:
Removed it.

lsiepel replied:
Removed this binding from this PR.
Created for reference: https://github.com/openhab/openhab-addons/issues/15507

jlaur commented on bundles/org.openhab.binding.energidataservice/src/main/java/org/openhab/binding/energidataservice/internal/ApiController.java:
```diff
@@ -110,7 +112,7 @@ public ElspotpriceRecord[] getSpotPrices(String priceArea, Currency currency, Da
 
         Request request = httpClient.newRequest(ENDPOINT + DATASET_PATH + DATASET_NAME_SPOT_PRICES)
-                .param("start", start.toString()) //
+                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MICROSECONDS).param("start", start.toString()) //
```
```suggestion
                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS) //
                .param("start", start.toString()) //
```

(or seconds, see other comment).

I would need to check average response time over some time to know if this is a reasonable timeout for this cloud service binding.

Usually the service is called only once per day at a very specific time. There is a retry policy with exponential backoff, so we'll retry automatically when hitting the timeout. However, it would be much preferable get it on first call to have the energy prices updated as fast as possible. Also, timeout is not part of the policy, so if it takes longer than 10 seconds, it might also take longer on next call etc.

I'm not saying it's bad, just that there are some considerations to make. I'll see if I can setup some metrics to get to know the min/avg/max response times.

lsiepel replied:
Let me know and i will adjust it. If at some point this PR is waiting for this, i can set it to 30seconds for now. And you might adjust it in a seperate PR.

jlaur replied:
Sure, 30 seconds would make me more calm and give me some time to measure and tweak.

lsiepel replied:
Done

jlaur commented on bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/GatewayWebTargets.java:
```diff
@@ -248,7 +249,8 @@ protected synchronized String invoke(HttpMethod method, String url, @Nullable Qu
         }
-        Request request = httpClient.newRequest(url).method(method).header("Connection", "close").accept("*/*");
+        Request request = httpClient.newRequest(url).method(method).header("Connection", "close").accept("*/*")
+                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MICROSECONDS);
```
```suggestion
                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```

lsiepel replied:
Fixed.

jlaur commented on bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/HDPowerViewWebTargets.java:
```diff
@@ -75,6 +76,7 @@
  */
 @NonNullByDefault
 public class HDPowerViewWebTargets {
+    private static final int REQUEST_TIMEOUT_MS = 10000;
```
I'm almost sure this value is too low. Some call will query the shades directly (i.e. HTTP call to hub -> RF to shade) for refreshing battery level, signal strength etc. Some of these calls are quite slow.

lsiepel replied:
Doubled it.

jlaur replied:
I just did a quick test of one specific API call I remember to be quite slow:
```
2023-08-26 23:40:11.059 [INFO ] [openhab.event.ItemCommandEvent      ] - Item 'Blind10_SignalStrength' received command REFRESH
2023-08-26 23:40:22.230 [INFO ] [rnal.handler.HDPowerViewShadeHandler] - No data from shade 24076 survey
```

I think 30 seconds would be on the safer side.

Slow calls are slow, and we'll have to wait for them. Or we would need to refactor to async calls. Fast calls shouldn't usually be a problem, even if timeout is 5 minutes, since they will only block for a very short time.

I'm now wondering if it would be better not to use the common HTTP client for sync call scenarios which are known to be slow - like this one and Bosch Indego.

lsiepel replied:
I'll set it to 30secs.
Don't think it will be a problem for the common client. But time will tell, if so we can create a PR easily.

jlaur commented on bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/HDPowerViewWebTargets.java:
```diff
@@ -581,7 +583,8 @@ private synchronized String invoke(HttpMethod method, String url, @Nullable Quer
         }
-        Request request = httpClient.newRequest(url).method(method).header("Connection", "close").accept("*/*");
+        Request request = httpClient.newRequest(url).method(method).header("Connection", "close").accept("*/*")
+                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MICROSECONDS);
```
```suggestion
                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```

lsiepel replied:
Fixed.

jlaur commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/MieleGatewayCommunicationController.java:
```diff
@@ -83,6 +85,7 @@ public JsonElement invokeRPC(String methodName, Object[] args) throws MieleRpcEx
 
         String requestBody = requestBodyAsJson.toString();
         Request request = httpClient.newRequest(uri).method(HttpMethod.POST)
+                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MICROSECONDS)
```
```suggestion
                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS)
```

(or better yet, use seconds)

lsiepel replied:
Fixed.

jlaur reviewed (changes requested):
I have provided some initial comments only for the bindings where I'm directly involved as code owner. This is a rather big change, and I can already see that there might be special considerations to make for individual bindings, as well as general considerations for best values for cloud vs. local API's.

andrewfg commented on bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartImpl.java:
```diff
@@ -212,6 +213,7 @@ private <T> T executeRequest(HttpMethod method, String url, @Nullable Object con
 
             Request request = httpClient.newRequest(url).method(method).header(HttpHeader.CONTENT_TYPE, contentType)
+                    .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MICROSECONDS)
                     .header(HttpHeader.ACCEPT, "application/vnd.api+json").header(HttpHeader.ACCEPT_ENCODING, "gzip");
```
Should be TimeUnit.MILLISECONDS

lsiepel replied:
Fixed.

daMihe reviewed (changes requested):
This timeout should be configurable. Using e.g. a System property etc. However this would make the change more complex.

lsiepel commented on bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/IndegoController.java:
```diff
@@ -119,6 +121,7 @@ protected <T> T getRequest(String path, Class<? extends T> dtoClass)
         int status = 0;
         try {
             Request request = httpClient.newRequest(BASE_URL + path).method(HttpMethod.GET)
+                    .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MICROSECONDS)
```
Doh.... Fixed. Everywhere.

lsiepel commented on bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/IndegoController.java:
```diff
@@ -59,6 +60,7 @@ public class IndegoController {
 
     private static final String BASE_URL = "https://api.indego-cloud.iot.bosch-si.com/api/v1/";
     private static final String CONTENT_TYPE_HEADER = "application/json";
+    private static final int REQUEST_TIMEOUT_MS = 10000;
```
Fixed. Everywhere.

mlobstein commented on bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/IndegoController.java:
```diff
@@ -12,19 +12,19 @@
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
-import org.eclipse.jetty.client.HttpClient;
```
I thought the use of the common Jetty HttpClient was required?

lsiepel replied:
Yeah needed some time to fix all the imports and grab a coffie. Fixed now.

mlobstein commented on bundles/org.openhab.binding.chatgpt/src/main/java/org/openhab/binding/chatgpt/internal/ChatGPTHandler.java:
```diff
@@ -31,7 +33,6 @@
 import org.openhab.binding.chatgpt.internal.dto.ChatResponse;
 import org.openhab.core.io.net.http.HttpClientFactory;
 import org.openhab.core.library.types.StringType;
-import org.openhab.core.thing.Channel;
```
Why should **org.openhab.core.thing.Channel** be swapped for **java.nio.channels.Channel**?

lsiepel replied:
VS Code did some nasty import stuff when editing files outside of the current folder. Fixed all of them.

mlobstein commented on bundles/org.openhab.binding.chatgpt/src/main/java/org/openhab/binding/chatgpt/internal/ChatGPTHandler.java:
```diff
@@ -40,7 +41,6 @@
 import org.openhab.core.thing.binding.ThingHandlerService;
 import org.openhab.core.types.Command;
 import org.openhab.core.types.RefreshType;
-import org.slf4j.Logger;
```
Seems like Logger is still needed below....

lsiepel replied:
Fixed.

mlobstein reviewed (changes requested):
Needs work...

jlaur commented on bundles/org.openhab.binding.energidataservice/src/main/java/org/openhab/binding/energidataservice/internal/ApiController.java:
```diff
@@ -76,6 +76,7 @@ public class ApiController {
 
     private static final String HEADER_REMAINING_CALLS = "RemainingCalls";
     private static final String HEADER_TOTAL_CALLS = "TotalCalls";
+    private static final int REQUEST_TIMEOUT_MS = 10_000;
```
Since the granularity of milliseconds is not used for this hardcoded constant, it could be expressed in seconds which would make it even more readable:
```suggestion
    private static final int REQUEST_TIMEOUT_SECONDS = 30;
```

lsiepel replied:
Done

jlaur commented on bundles/org.openhab.binding.energidataservice/src/main/java/org/openhab/binding/energidataservice/internal/ApiController.java:
```diff
@@ -110,7 +111,7 @@ public ElspotpriceRecord[] getSpotPrices(String priceArea, Currency currency, Da
 
         Request request = httpClient.newRequest(ENDPOINT + DATASET_PATH + DATASET_NAME_SPOT_PRICES)
-                .param("start", start.toString()) //
+                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS).param("start", start.toString()) //
```
Formatting and unit commented above.
```suggestion
                .timeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS) //
                .param("start", start.toString()) //
```

lsiepel replied:
Done

jlaur commented on bundles/org.openhab.binding.energidataservice/src/main/java/org/openhab/binding/energidataservice/internal/ApiController.java:
```diff
@@ -198,7 +199,7 @@ public Collection<DatahubPricelistRecord> getDatahubPriceLists(GlobalLocationNum
 
         Request request = httpClient.newRequest(ENDPOINT + DATASET_PATH + DATASET_NAME_DATAHUB_PRICELIST)
-                .param("filter", mapToFilter(filterMap)) //
+                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS).param("filter", mapToFilter(filterMap)) //
```
Formatting and unit commented above.
```suggestion
                .timeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS) //
                .param("filter", mapToFilter(filterMap)) //
```

lsiepel replied:
Done

mlobstein commented on bundles/org.openhab.binding.radiothermostat/src/main/java/org/openhab/binding/radiothermostat/internal/communication/RadioThermostatConnector.java:
```diff
@@ -124,7 +125,8 @@ public String sendCommand(@Nullable String cmdKey, @Nullable String cmdVal, @Nul
 
         try {
-            Request request = httpClient.POST(buildRequestURL(resource));
+            Request request = httpClient.POST(buildRequestURL(resource)).timeout(REQUEST_TIMEOUT_MS,
+                    TimeUnit.MICROSECONDS);
```
```suggestion
            Request request = httpClient.POST(buildRequestURL(resource)).timeout(REQUEST_TIMEOUT_MS,
                    TimeUnit.MILLISECONDS);
```

lsiepel replied:
Thanks, fixed

lolodomo replied:
Not yet fixed.

lsiepel replied:
Weird, don't know what happened, anyway fixed.

mlobstein commented on bundles/org.openhab.binding.tellstick/src/main/java/org/openhab/binding/tellstick/internal/local/TelldusLocalDeviceController.java:
```diff
@@ -270,7 +272,8 @@ private <T> T innerCallRest(String uri, Class<T> json)
         logger.trace("HTTP GET: {}", uri);
 
-        Request request = httpClient.newRequest(uri).method(HttpMethod.GET);
+        Request request = httpClient.newRequest(uri).method(HttpMethod.GET).timeout(REQUEST_TIMEOUT_MS,
+                TimeUnit.MICROSECONDS);
```
```suggestion
        Request request = httpClient.newRequest(uri).method(HttpMethod.GET).timeout(REQUEST_TIMEOUT_MS,
                TimeUnit.MILLISECONDS);
```

lsiepel replied:
Thanks, fixed

lolodomo replied:
Not yet fixed.

lsiepel replied:
Weird, don't know what happened, anyway fixed.

mlobstein reviewed (changes requested):
A couple more MICROSECONDS...

lsiepel commented:
> This timeout should be configurable. Using e.g. a System property etc. However this would make the change more complex.

I’m not planning to make it configurable for each binding. If needed, a seperate follow up PR can be created for that binding.

mlobstein reviewed (approved):
LGTM

lsiepel commented:
Build error seems unrelated:
https://ci.openhab.org/job/PR-openHAB-Addons/org.openhab.addons.bundles$org.openhab.binding.groupepsa/16918/testReport/org.openhab.binding.groupepsa.internal.things/GroupePSAHandlerTest/

dougculnane reviewed (approved):
Looks good to me.  10 seconds should be enough for the Renault API requests.

lolodomo commented on bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/FreeboxOsIconProvider.java:
```diff
@@ -51,6 +52,7 @@
 public class FreeboxOsIconProvider extends AbstractResourceIconProvider {
 
     private final Logger logger = LoggerFactory.getLogger(FreeboxOsIconProvider.class);
+    private static final int REQUEST_TIMEOUT_MS = 8000;
```
By curiosity, why this value ?

@clinique : is a reasonable value in your opinion ?

lsiepel replied:
Because of the 'business rule' i mentioned in the startpost. Elsewhere this binding has timeouts set to 8 seconds.
https://github.com/openhab/openhab-addons/blob/c5739eccc978f520473750871074d1e94a7579d5/bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/ApiHandler.java#L69

lolodomo commented:
@lsiepel : there is no default timeout aplied when not set ?

lsiepel commented:
> @lsiepel : there is no default timeout applied when not set ?

There does not seem to be a default documented.

jlaur commented:
> there is no default timeout applied when not set ?

I was recently revising some old stashed changes for the Bosch Indego binding when I was trying to implement **longpoll** (see #13182). I had configured a custom thread pool, custom http client and set a long timeout on the request created. But yet, I had `EOFException` after exactly 5 minutes.

So from those experiments it seems there is actually a default timeout of 5 minutes. @lsiepel, what is your experience here? From #11527 which started this, it was mentioned that http client would "block forever". So I'm a little confused about some details here about the actual issue.

lolodomo commented:
But even with 5 minutes, it would be largely too much. So setting a specific timeout would be better.

jlaur commented:
> But even with 5 minutes, it would be largely too much. So setting a specific timeout would be better.

Yes, I agree. But maybe there is some way to configure the default timeout in Jetty, so we could provide a more reasonable value? I mostly wanted express my confusion about the "blocking forever" problem statement that is being addressed, since I observed a different outcome: 5 minutes.

lsiepel commented:
> > But even with 5 minutes, it would be largely too much. So setting a specific timeout would be better.
> 
> Yes, I agree. But maybe there is some way to configure the default timeout in Jetty, so we could provide a more reasonable value? This would also allow us to change this default without hardcoding it into all bindings.
> 
> I mostly wanted express my confusion about the "blocking forever" problem statement that is being addressed, since I observed a different outcome: 5 minutes.

5minutes is like forever ;-) 
I don't know the code you are referring to, but if the request timeout defaults to 5 minutes, i would expect a `TimeOutException`, not a `EOFException`. From my understanding of that issue, setting the timeout to 5+minutes also generates the same exception at 5 minute mark. So that also points to a different issue as the explicit set timeout is ignored. 
I really don't think there is a default.

Anyway. Let's agree that setting an explicit timeout is _allways_ a best practice. So i'm happy to adjust any timeout here. I used some rule upfront to determine the right timeout, let me know if i have to change that.

lsiepel commented:
Ah ffs, i messed up the merge. I need some help to get this back together.

lsiepel commented:
@jlaur @lolodomo sorry for all the fuzz. I still maintain a hate/love relation with git and vscode. Anyway, the merge conflict is solved. I hope you can merge this so it can get some mileage on milestones. If needed i can further adjust timeouts.

marcelrv commented:
LGTM, but don't think any of my bindings is impacted.

lolodomo commented:
DCO is KO.

lsiepel commented:
> DCO is KO.

KO?

daMihe commented on bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/PullJob.java:
Does it make sense to add a timeout if it is already enforced through it's listener? 

If yes: please use the defined `HTTP_TIMEOUT_SECS`.

(Just reviewed icalendar)

lsiepel replied:
Thanks, fixed.

daMihe replied:
LGTM for icalendar.

daMihe reviewed (approved):
LGTM for icalendar

lsiepel commented:
Not sure what is wrong with the build. Labals also don;t seem to work. Looks unrelated to this pr.

lolodomo commented:
@clinique : could we have your approval for freeboxos binding please ?

@jlaur : the general goal of this PR is fine IMHO. I see you were involved in its review. Did you review it fully ? In this case, I let you conclude.

I am just annoyed that there is commit probably not well signed.

jlaur commented:
> the general goal of this PR is fine IMHO. I see you were involved in its review. Did you review it fully ? In this case, I let you conclude.

I have not reviewed fully, but all my comments for specific bindings I'm "protecting" as code owner were considered.

clinique reviewed (approved):
LGTM

lsiepel commented:
Would be nice to have this merged before the next milestone. I don't expect any issue, but if there is a binding that needs further adjustment we can then fix it before stable release. Adjusting these timeouts is very simple.

lolodomo commented on bundles/org.openhab.binding.liquidcheck/src/main/java/org/openhab/binding/liquidcheck/internal/httpclient/LiquidCheckHttpClient.java:
```diff
@@ -62,7 +62,8 @@ public LiquidCheckHttpClient(LiquidCheckConfiguration config, HttpClient client)
         String uri = "http://" + config.hostname + "/infos.json";
         Request request = client.newRequest(uri).method(HttpMethod.GET)
-                .timeout(config.connectionTimeout, TimeUnit.SECONDS).followRedirects(false);
+                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS).timeout(config.connectionTimeout, TimeUnit.SECONDS)
```
A timeout was already defined here, you now defined a doubled timeouts.

lolodomo replied:
You could keep the original timeout here

lsiepel replied:
reverted to the original

lolodomo commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -416,8 +418,8 @@ private synchronized ContentResponse sendRequest(String url, HttpMethod method,
-        Request request = httpClient.newRequest(url).method(method).agent(userAgent).timeout(10, TimeUnit.SECONDS)
-                .header("Authorization", authTokenHeader(tokenResponse));
+        Request request = httpClient.newRequest(url).method(method).agent(userAgent)
+                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.SECONDS).header("Authorization", authTokenHeader(tokenResponse));
```
Wrong unit

lsiepel replied:
Unit fixed

lolodomo commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -528,6 +531,7 @@ private String postLogin(String url, String requestToken, String returnURL)
         fields.add("ReturnUrl", returnURL);
 
         Request request = httpClient.newRequest(url).method(HttpMethod.POST) //
+                .timeout(REQUEST_TIMEOUT_MS, TimeUnit.SECONDS) //
```
Wrong unit

lsiepel replied:
Fixed unit

lolodomo commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -554,7 +558,8 @@ private String postLogin(String url, String requestToken, String returnURL)
                 break;
             }
-            request = httpClient.newRequest(LOGIN_BASE_URL + loc).agent(userAgent).followRedirects(false);
+            request = httpClient.newRequest(LOGIN_BASE_URL + loc).timeout(REQUEST_TIMEOUT_MS, TimeUnit.SECONDS)
```
Wrong unit

lsiepel replied:
Fixed unit

lolodomo commented:
> Would be nice to have this merged before the next milestone. I don't expect any issue, but if there is a binding that needs further adjustment we can then fix it before stable release. Adjusting these timeouts is very simple.

I have done a full review and discovered things to fix.
And we have a problem with DCO, one of your commit is probably not properly signed-off.

lsiepel commented:
> > Would be nice to have this merged before the next milestone. I don't expect any issue, but if there is a binding that needs further adjustment we can then fix it before stable release. Adjusting these timeouts is very simple.
> 
> I have done a full review and discovered things to fix. And we have a problem with DCO, one of your commit is probably not properly signed-off and this will not be possible to merge the PR considering its size and how many bindings are impacted.

Fixed the comments, thanks!

I don't see how the DCO can be a problem for this case. I messed up the merge and reverted that. The revert commit (not adding any code) is not signed, due to github desktop. All other commits (adding or changing code) are signed.
If you know a quick way for me to fix this let me know, buit i expect that i make it worse :-/

andrewfg commented:
> I don't see how the DCO can be a problem for this case. 

@lsiepel if you click on Details next to the DCO check, it takes you to a page where you can manually "Set DCO to Pass" :)

lsiepel commented:
Build error is unrelated: https://github.com/openhab/openhab-addons/issues/15646

weymann reviewed (approved):
LGTM for Mercedes

lolodomo reviewed (approved):
LGTM, thank you

lolodomo commented:
@lsiepel : please fix the conflict so that I can merge.
